components/proxy/src/main/kotlin/com/hotels/styx/routing/db/StyxObjectStore.kt

```diff
@@ -15,28 +15,100 @@
  */
 package com.hotels.styx.routing.db;
 
-import com.hotels.styx.api.configuration.ObjectStore
+import com.hotels.styx.api.configuration.ObjectStoreSnapshot
+import org.pcollections.HashTreePMap
+import org.pcollections.PMap
+import org.reactivestreams.Publisher
+import reactor.core.publisher.Flux
 import java.util.Optional
-import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Consumer
 
 /**
  * Styx Route Database.
  */
-class StyxObjectStore<T> : ObjectStore<T> {
-    private val objects = ConcurrentHashMap<String, Record<T>>();
+class StyxObjectStore<T> : ObjectStoreSnapshot<T> {
+    private val objects: AtomicReference<PMap<String, Record<T>>> = AtomicReference(HashTreePMap.empty())
+    private val watchers = CopyOnWriteArrayList<ChangeWatcher<T>>()
+    private val executor = Executors.newSingleThreadExecutor()
 
-    override fun get(name: String?): Optional<T> {
-        return Optional.ofNullable(objects.get(name))
+    override fun get(name: String): Optional<T> {
+        return Optional.ofNullable(objects().get(name))
                 .map { it.payload }
     }
 
+    fun insert(key: String, payload: T) {
+        insert(key, setOf(), payload)
+    }
+
+    fun remove(key: T) {
+        queue {
+            val nextVersion = objects().minus(key)
+            objects.set(nextVersion)
+            notifyWatchers(nextVersion)
+        }
+    }
+
+    /**
+     * Returns a Publisher that emits an event at any modification.
+     *
+     * Watch activates on subscription only.
+     * Watch removed on unsubscription.
+     */
+    fun watch(): Publisher<ObjectStoreSnapshot<T>> {
+        return Flux.push { sink ->
+            val watcher = object : ChangeWatcher<T> {
```

Comment:
Can't we simplify this?

Reply:
Looks like we can. By using Kotlin type alias instead of Java interfaces. Given:

    private typealias ChangeWatcher<T> = (ObjectStore<T>) -> Unit

Then:

    val watcher: ChangeWatcher<T> = { sink.next(it) }

Reply:
Pretty nice :)